### ISSUE TITLE
Runtime conflict with druntime

### ISSUE THREAD

**Issue by jacob-carlborg** [SiegeLord/Tango-D2#47]
There are some conflicts between the functionality in `tango.core.Runtime` and `core.runtime`. For example, the unit test runner in Tango seems to override the one in druntime.

BTW why are some parts of the Tango runtime left, like:

`tango.core.` -
- Runtime
- Thread
- Vararg
- Memory

**Comment by SiegeLord** [comment]
Those are backwards compatibility modules.

I am confused about the nature of the bug here... is this an issue of using both in the same program? Or the Tango one not working at all?

**Comment by jacob-carlborg** [comment]
The issue is the following:

``` d
module main;
import std.stdio;

unittest {
    writeln("foo");
}

void main () { }
```

```
$ dmd main.d -unittest -L-ltango-d2 && ./main
```

The expected output is "foo", but there is no output. As soon as I link with Tango it won't run the unit tests. Note that I don't even need to use Tango in the application, I just need to link with Tango. I think the issue is [tango.core.Runtime.runModuleUnitTests](https://github.com/SiegeLord/Tango-D2/blob/d2port/tango/core/Runtime.d#L169). druntime contains a function with the exact same name. Since both the one in Tango and the one in druntime is declared as `extern(C)` they will have the same name. I think the one in Tango overrides the on in druntime.

The version of `runModuleUnitTests` in Tango would work fine, expect that [ModuleInfo.unitTest](https://github.com/D-Programming-Language/druntime/blob/master/src/object_.d#L1753) is a method that returns a function pointer these days. If you look a the implementation of `runModuleUnitTests` in Tango you can see that it calls `m.unitTest()` to run a given unit test block. Due to how properties work in D this will only call the `unitTest` method and not the function pointer returning it.

I think these modules should be removed from Tango, they're part of the runtime, which druntime implements now. If these exist along side with the ones in druntime Tango won't be compatible with druntime and we have the same problem again as we did with D1.

**Comment by SiegeLord** [comment]
I can't reproduce this behavior (the code you pasted runs just fine for me). What OS is this?

I'll see what I can do about the mismatched function types. I'd rather not alias tango.core.Runtime with core.runtime since Tango's version of that module has features missing in druntime... I'll see if I can combine them somehow.

**Comment by jacob-carlborg** [comment]
> I can't reproduce this behavior (the code you pasted runs just fine for me). What OS is this?

Mac OS X.

> I'll see what I can do about the mismatched function types. I'd rather not alias tango.core.Runtime with core.runtime since Tango's version of that module has features missing in druntime... I'll see if I can combine them somehow.

That would be nice.

**Comment by jacob-carlborg** [comment]
The command DMD uses to link is:

```
gcc main.o -o main -m64 -ltango -L/Users/jacob/.dvm/compilers/dmd-2.060/bin/../lib -lphobos2 -lpthread -lm
```

If I move `-ltango` to come after `-lphobos2` everything works as expected.

**Comment by jacob-carlborg** [comment]
A guess, remove `extern(C)` from `runModuleUnitTests` and it will work.

**Comment by SiegeLord** [comment]
Could you try it now, now that 6a290126eda6e6cb7fbeaaa5b6b697ca4f463180 is committed?

**Comment by jacob-carlborg** [comment]
Thanks that works. Fixed in 6a290126eda6e6cb7fbeaaa5b6b697ca4f463180
